### PR TITLE
Update `rustls` to 0.20.1

### LINKED ISSUE #644: upgrade rustls, pem, tame-oauth, and webpki
These have breaking changes that were found during the last version upgrade in #643.

The relevant backouts that we would like to undo are: https://github.com/kube-rs/kube-rs/pull/643/commits/87eeb01130bbd1a972b8148620f2520651366035 and https://github.com/kube-rs/kube-rs/pull/643/commits/f2890b1e07e1948130e5f469da30aa7a0d371424 .

- [x] [breaking changes in rustls](https://github.com/rustls/rustls#release-history) - #692, #704
  - [x] Remove `webpki`
- [x] [pem changelog for 1.0](https://github.com/jcreekmore/pem-rs/blob/master/CHANGELOG.md) - #702
- [x] [tame-oauth changelog for 04.7 to 0.6.0](https://github.com/EmbarkStudios/tame-oauth/blob/main/CHANGELOG.md) - #701

Also likely crypto related: we bumped `tokio-tungstenite` by one minor to [0.15.0](https://github.com/snapview/tokio-tungstenite/blob/master/CHANGELOG.md#0150), giving us `tungstenite-rs`[@0.14.0](https://github.com/snapview/tungstenite-rs/blob/master/CHANGELOG.md#0140) (with some `ring` bump changes).

### PR DESCRIPTION
**Pull request by kazk**
Continuing from #692 for #644 after rebasing. `webpki` is no longer necessary, so this is the last one.

#### TODO

- [x] Refine errors
- [x] Replace `hyper_rustls::HttpsConnector::with_native_roots` with `HttpsConnectorBuilder`
- [x] Use published `hyper-rustls` (https://github.com/rustls/hyper-rustls/pull/159)

---

Closes #644

### CODE REVIEW

File: kube-client/src/client/tls.rs
Comment (kazk) [inline kube-client/src/client/tls.rs:119]: Changed to get all sections at once, but it should behave the same. Use the last found private key, and PKCS8 key is preferred.

File: examples/pod_watcher.rs
Comment (kazk) [inline examples/pod_watcher.rs:16]: @clux Should I drop this change?

Comment (clux) [reply examples/pod_watcher.rs:16]: ah, whoops, yes please. that was a debug for me while developing.

File: kube-client/src/client/tls.rs
Comment (clux) [inline kube-client/src/client/tls.rs:104]: client auth method simplified a lot :+1:

**Comment by clux** [comment]
This seems to work as well as before. Thought I had a regression on rustls vs rancher, but turns out rancher just never worked with local kubeconfig on rustls.

rancher gives this type of setup:

```yaml
clusters:
- cluster:
    server: https://rancher.xxx.io/k8s/clusters/clusterid
  name: mycluster
users:
- name: myuser
  user:
    token: kubeconfig-u-somegenuser.c-id:somelongtoken
contexts:
- context:
    cluster: mycluster
    user: myuser
  name: mycontext
```

<details><summary>error</summary>
<p>

```sh
[2021-11-15T22:06:29Z DEBUG kube_client::config] failed to load client identity from kubeconfig: failed to load client certificate 
[2021-11-15T22:06:29Z DEBUG kube_client::client] HTTP; http.method=GET http.url=https://rancher.xxx.io/k8s/clusters/clusterid/api/v1/namespaces/default/pods? otel.name="list" otel.kind="client"
[2021-11-15T22:06:29Z DEBUG kube_client::client] requesting 
[2021-11-15T22:06:30Z WARN  rustls::conn] Sending fatal alert BadCertificate
[2021-11-15T22:06:30Z DEBUG kube_client::client] HTTP; otel.status_code="ERROR"
[2021-11-15T22:06:30Z ERROR kube_client::client] failed with error error trying to connect: invalid peer certificate contents: invalid peer certificate: UnknownIssuer 
Error: failed to perform initial object list: HyperError: error trying to connect: invalid peer certificate contents: invalid peer certificate: UnknownIssuer

Caused by:
   0: HyperError: error trying to connect: invalid peer certificate contents: invalid peer certificate: UnknownIssuer
   1: error trying to connect: invalid peer certificate contents: invalid peer certificate: UnknownIssuer
   2: invalid peer certificate contents: invalid peer certificate: UnknownIssuer

Location:
    /rustc/59eed8a2aac0230a8b53e89d4e99d55912ba6b35/library/core/src/result.rs:1915:27

```

</p>
</details>

interestingly this works with openssl, but with rustls i have to add `insecure-skip-tls-verify: true` to the cluster entry. not sure if you have an idea of this. can open a separate bug if not.

**Comment by kazk** [comment]
> interestingly this works with openssl, but with rustls i have to add `insecure-skip-tls-verify: true` to the cluster entry. not sure if you have an idea of this. can open a separate bug if not.

I think `native-tls` includes the system cert store by default, while `rustls` doesn't. We can load with `rustls-native-certs` which is already a dependency of `hyper-rustls`.

File: kube-client/src/client/tls.rs
Comment (kazk) [inline kube-client/src/client/tls.rs:100]: Loads the native roots.

@clux Can you try Rancher again?

Comment (clux) [reply kube-client/src/client/tls.rs:100]: Just checked. It's fixed!

**Comment by kazk** [comment]
#711 is working :)
```text
error[B004]: found 2 duplicate entries for crate 'rustls'
    ┌─ /github/workspace/Cargo.lock:151:1
    │  
151 │ ╭ rustls 0.19.1 registry+rust-lang/crates.io-index
152 │ │ rustls 0.20.1 registry+rust-lang/crates.io-index
    │ ╰───────────────────────────────────────────────────────────────────^ lock entries
    │  
    = rustls v0.19.1
      └── tokio-rustls v0.22.0
          └── warp v0.3.2
              └── (dev) examples v0.1.0
    = rustls v0.20.1
      ├── hyper-rustls v0.23.0
      │   └── kube-client v0.63.2
```

File: kube-client/src/error.rs
Comment (clux) [inline kube-client/src/error.rs:76]: think you need to:

```suggestion
    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-tls")))]
```

got a rustdoc error from a similar thing in the previous pr

Comment (kazk) [reply kube-client/src/error.rs:76]: Ah, thanks. I probably copied and pasted what you fixed in that PR :|

**Review by clux — APPROVED** [review]
Great work :+1:

**Comment by kazk** [comment]
I'll merge this after cleaning up the commits:

- Move all edits to `Cargo.toml` to the first commit
- Merge commit removing `Error::SslError` with refining error one

I thought of squash merging, but I think it's better to keep the upgrade and refactoring separate after merging.

**Pull request merged**
